astropy/table/table.py

```diff
@@ -3674,3 +3683,42 @@ def __setstate__(self, state):
         nd_state, own_state = state
         super().__setstate__(nd_state)
         self.__dict__.update(own_state)
+
+
+class TableAttribute:
+    """
+    Descriptor to define custom Table attribute which gets stored in the object
+    meta dict and can have a defined default.
+
+    :param default: default value
+
+    """
+    def __init__(self, default=None):
+        self.default = default
+
+    def __get__(self, instance, owner):
+        try:
+            return instance.meta[self.name]
+        except AttributeError:
+            # When called without an instance, return self to allow access
+            # to descriptor attributes.
+            # AttributeError: 'NoneType' object has no attribute 'meta'
+            return self
+        except KeyError:
+            if self.default is not None:
+                instance.meta[self.name] = deepcopy(self.default)
+            return instance.meta.get(self.name)
+
+    def __set__(self, instance, value):
+        instance.meta[self.name] = value
+
+    def __set_name__(self, owner, name):
+        reserved = ('data', 'masked', 'names', 'dtype',
+                    'meta', 'copy', 'rows', 'copy_indices',
+                    'units', 'descriptions')
+        if hasattr(Table, name) or name in reserved:
```

Comment:
Shouldn't this be `hasattr(owner, name)`? If the subclass defines a new method, presumably that shouldn't be overwritten either.

More because I'm a bit confused: the `reserved` set look like the keywords in `__init__` - maybe get it automatically by inspection? Or is it in fact needed? (`meta` and `copy` would not seem needed, since they exist already).

Reply:
Yes, this should be `owner` not hard-coded Table.  I'm thinking now about `reserved` as a class attribute of `MetaAttribute`, so users can subclass.  Then perhaps we still end up with `TableAttribute` as a subclass of `MetaAttribute`.

The idea of `reserved` is to explicitly avoid collisions in the class `__init__` that might cause confusion.  Especially if this is put out there as something to use on any class, then there is really no way to *reliably* infer the API even with inspect for classes that hav a catch-all `**kwargs`.

Reply:
This perhaps is an argument to stick with `TableAttribute` here for now, and give ourselves the flexibility to implement a more general `MetaAttribute` later, from which `TableAttribute` can then inherit.